docs/reference/xpkg.md

```diff
@@ -0,0 +1,201 @@
+> the controller if the package is a Provider. The decision of selecting a
+> platform for a Provider controller image is deferred to the configured
+> container runtime.
+
+### Manifests
+
+A manifest defines the layers and configuration of a specific image. Crossplane
+is only concerned with the layer descriptors array in an image manifest and does
+not impose additional requirements on any other portions of the manifest. The
+following requirements are imposed on the layer descriptors array:
+
+- One (1) layer descriptor in the array MAY have an [annotation] with key
+  `io.crossplane.xpkg` and value `base`.
+- Any number of layer descriptors in the array MAY have an annotation with key
+  `io.crossplane.xpkg` and arbitrary value. Whether multiple layer descriptors
+  may have the same value is left to the specification of the consumer of those
+  layers.
+
+> As evidenced by the fact that annotations are provided as a map of
+> _string-string_, no single descriptor will contain multiple
+> `io.crossplane.xpkg` annotations.
+
+Crossplane is only concerned with the layer with the `base` annotation, and any
+other layers with the `io.crossplane.xpkg` key are used to signify to
+third-party consumers that a layer contains content related to the `xpkg` that
+may be specific to a given consumer.
+
+If no layer descriptors have an annotation in the form `io.crossplane.xpkg:
+base`, the resultant filesystem from [applying changesets] from all layers will
+be used.
+
+> Crossplane prefers the usage of layers to define additive package content as
+> it provides a clean mechanism to build an `xpkg` through a series of stages. A
+> valid `xpkg` can be produced and later modified while verifying that the
+> integrity of the existing content is not violated, which ensures that
+> Crossplane's package manager will process the resulting `xpkg` in the same
+> manner as the it would prior to modification.
+
+### Configuration
+
+Crossplane imposes no additional requirements on image configuration and does
+not consider its contents when processing a package.
+
+### Layers
+
+As described above, Crossplane is only concerned with the single layer
+referenced by the descriptor containing `io.crossplane.xpkg: base` if
+distinguished. Crossplane imposes no additional restrictions on any other
+layers, including those with a `io.crossplane.xpkg` annotation but a value other
+than `base`, but does require the following of the `xpkg` base layer:
+
+- A single file with name `package.yaml` MUST exist in the root directory of the
+  `xpkg` base layer if distinguished, or in the root of the image filesystem
+  after all layer changesets are applied.
+- The `package.yaml` file MUST contain a valid [YAML stream].
+- All other content in either the `xpkg` base layer, or the full image
+  filesystem is ignored by Crossplane.
+
+> The ability to use the image's flattened filesystem is primarily for backwards
```

Comment:
What do we gain by taking a specific layer into consideration instead of using the resulting filesystem all the time? We've got the tools to work with a specific layer but treating the image as a single filesystem is easier for people to reason about and work with, like just `docker run` and copy something etc.

Reply:
@muvaf I think this is called out in the following sentences, but the things we gain include:
- If subsequent layers overwrite content the package is still valid. This means an if image is built incrementally, there is less chance of ending up with an image that does not work as a Crossplane package. While it may seem like this is unlikely, in the event the `package.yaml` in the root is overwritten it would be somewhat difficult to debug, and if you are using extensions there is potentially "3rd party package content" outside of the `package.yaml`.
- Specifying a single layer means that we don't have to download the entire image to parse the package contents. This becomes especially useful if we move to using combined controller / package images more frequently, and/or with using the extension mechanisms described.

It is also worth pointing out that this specification explicitly allows for _not_ specifying a single layer, which means the path that you are talking about will still work as expected. For the reasons mentioned above, using the flattened filesystem approach may result in more inefficient parsing (only at the beginning, we can opt to only cache the package.yaml content).

Reply:
Could we advise, or require, that the `package.yaml` is never mutated or removed in subsequent layers? I like that we can keep it in an easily identified layer for downloads and caching, but would prefer to make sure it's always visible there in the resulting FS to avoid potentially surprising users.

Reply:
> Specifying a single layer means that we don't have to download the entire image to parse the package contents.

Thank you for explaining, makes a lot of sense to me.

What do you think about having a `Motivation` section to put those reasonings so that if people discuss changing it, they can have a clearer picture of why we do this?

Reply:
> Could we advise, or require, that the package.yaml is never mutated or removed in subsequent layers?

@negz I think I would push back on requiring that, and I think this gets to the broader purpose of this spec: defining what a valid package is, rather than how to build one. From that perspective, mandating that the `package.yaml` not be changed would mean that we would need to reject a package in Crossplane in the following scenario:
- None of the layers are annotated
- There are multiple layers
- The flattened file system contains a valid `package.yaml` at root
- The `package.yaml` was touched in multiple layers

Actually determining that the `package.yaml` was touched in multiple layers would be a fairly cumbersome process that I don't see providing a lot of value. I would advise customers to use tooling that doesn't allow that to happen, and I think any tooling in the Crossplane community should take that approach, but one of the reasons that we are maintaining the ability to use the flattened filesystem at all is that folks may use something like Docker to build their packages and mandating that the `package.yaml` is put into that filesystem in a certain manner makes that process more complicated IMO.

I do think warning against potentially unexpected behavior in documentation on building packages makes a lot of sense though 👍🏻 

> What do you think about having a Motivation section to put those reasonings so that if people discuss changing it, they can have a clearer picture of why we do this?

@muvaf great idea! Will add 👍🏻

Reply:
@hasheddan I'm concerned about it being able to do things that seem clever/surprising to those (presumably majority of) folks who don't have as deep a knowledge of OCI images as yourself. I would certainly find it surprising if I were to inspect a Crossplane package naively (i.e. by looking at the 'complete' set of layers) and find that either:

* There was no `package.yaml`.
* There was a `package.yaml` that was different than what Crossplane was actually using.

Perhaps the requirement should be that the `package.yaml` in the annotated layer must match the 'final' `package.yaml` in the flattened filesystem?

I feel comfortable about having the spec say not to do something without necessarily writing code to enforce that the thing is not done - especially within the actual package manager (as opposed to a linter where having to have all the layers in order to validate it seems less onerous).

Reply:
> I feel comfortable about having the spec say not to do something without necessarily writing code to enforce that the thing is not done

I don't think I am aligned here as it could be viewed as Crossplane not adhering to its own spec if there is explicit behavior defined that is not honored. I would be surprised as a user if a tool didn't let me build something because it was adhering to the spec, but Crossplane would let me install it. Also keep in mind the following points:
- If someone is less familiar with OCI images and they are using a tool like Docker, they likely are not annotating a layer and the `package.yaml` in the flattened filesystem will be what is used by default.
- If someone is using Crossplane tooling that annotated the layer then they likely are either building a fresh image that is only a package (i.e. single layer) or they are modifying an existing image. In the former case there is no chance for the `package.yaml` to be altered so it will be reflected in flattened filesystem (which is the same as the single layer filesystem). In the latter case, the tooling can determine exactly how it wants to add the `package.yaml`, potentially being more specific than Crossplane requires.

With this in mind, I would prefer for the spec to be exact about what will be accepted as a valid package, and allow users to dictate the UX they want for building.

Reply:
After some offline discussion I am going to keep the spec as-is, but will add a note mentioning that mutating the `package.yaml` in subsequent layers could lead to confusion for users. Crossplane will not explicitly disallow it (i.e. packages will not be rejected if they are valid according to the requirements outlines in this document).

Thanks for the discussion on this one @negz and @muvaf!